be2-scala/src/main/scala/ch/epfl/pop/json/HighLevelProtocol.scala

```diff
@@ -198,6 +199,36 @@ object HighLevelProtocol extends DefaultJsonProtocol {
     }
   }
 
+  implicit object RumorFormat extends RootJsonFormat[Rumor] {
+    final private val PARAM_SENDER_PK: String = "sender_id"
+    final private val PARAM_RUMOR_ID: String = "rumor_id"
+    final private val PARAM_MESSAGES: String = "messages"
+
+    override def read(json: JsValue): Rumor = {
+      val jsonObject: JsObject = json.asJsObject
+      jsonObject.getFields(PARAM_SENDER_PK, PARAM_RUMOR_ID, PARAM_MESSAGES) match {
+        case Seq(senderPk @ JsString(_), rumorId @ JsNumber(_), rumors @ JsObject(_)) =>
+          val map = mutable.HashMap[Channel, Array[Message]]()
```

Comment:
I think you can reduce the complexity of this, especially the use of a mutable HashMap
Try:
val map: Map[Channel, Array[Message]] = rumors.asJsObject.fields.map {
  case (k: String, JsArray(v)) => Channel(k) -> v.map(_.convertTo[Message]).toArray
  case _                       => throw new IllegalArgumentException(s"Unrecognizable rumor in $json")
}

Tested it and it passes the tests